File: golang/binary-tree-general/binary_search_tree_iterator.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type BSTIterator struct {
	stack []*TreeNode
}

func Constructor(root *TreeNode) BSTIterator {
	iterator := BSTIterator{}
	iterator.leftmostInorder(root)

	return iterator
}

func (this *BSTIterator) leftmostInorder(node *TreeNode) {
	for node != nil {
		this.stack = append(this.stack, node)

		node = node.Left
	}
}

func (this *BSTIterator) Next() int {
	topNode := this.stack[len(this.stack)-1]
	this.stack = this.stack[:len(this.stack)-1]

	if topNode.Right != nil {
		this.leftmostInorder(topNode.Right)
	}

	return topNode.Val
}

func (this *BSTIterator) HasNext() bool {
	return len(this.stack) > 0
}

File: python/binary-tree-general/binary_search_tree_iterator.py
from typing import Optional


class TreeNode:
    def __init__(self, val:int = 0, left = None, right = None) -> None:
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root: Optional[TreeNode]) -> None:
        self.stack = []
        self._left_most_inorder(root)

    def _left_most_inorder(self, node):
        while node:
            self.stack.append(node)

            node = node.left

    def next(self) -> int:
        top_node = self.stack.pop()

        if top_node.right:
            self._left_most_inorder(top_node.right)

        return top_node.val

    def hasNext(self) -> bool:
        return len(self.stack) > 0
